nntrainer/tensor/var_grad.h

```diff
@@ -346,7 +356,13 @@ class Var_Grad {
    * @note this is can return is the var_grad needs gradient but it not
    * empty
    */
-  bool hasGradient() const { return !grad->empty(); }
+  bool hasGradient() const {
+    if (!grad)
+      return false;
+    if (var->isAllocated())
+      return grad->isAllocated();
```

Comment:
`hasGradient()` combines the below questions into one:
- was this var_grad created with gradient
- does the gradient has defined shape
- has the gradient been allocated

There are various scenarios where its used with different concepts, and it would require all these checks independently to be done here. This function wraps all of them, and provides guarantees that if the return value is true, gradient is available and can be used safely.

Reply:
> > > We can possibly remove setBatch(RunLayerContext)
> > 
> > 
> > 
> > 1. Can we remove setBatch(RunLayerContext) ? this is used for the supporting setting batch after the initialization.
> > 2. Can we remove setBatch(InitLayerContext) ? My answer is yes but seeing this PR, maybe there might be another usage.
> 
> The purpose of either of them is to able to allow the user to update the batchsize of the tensors they have requested.
> 
> 1. setBatch(RunLayerContext) - this is the easier choice from implementation perspective
> 2. setBatch(InitLayerContext) - this is easier for the user IMO as user used InitLayerContext to request the tensors in the first place
> 
> But only one is needed. So, the other will be removed soon.

Got the gist. Thanks.